CMakeLists.txt

```diff
@@ -0,0 +1,481 @@
+    ./Client/*Mock.cc
+    ./Storage/*Mock.cc
+    ./Server/*Mock.cc
+    )
+
+## search for example files
+file(GLOB EXAMPLE_FILES 
+    ./Examples/*.cc
+    )
+
+# protoc generator
+
+macro(c_proto_buff INPUT_LIST OUTPUT_LIST)
+    list(LENGTH ${INPUT_LIST} LIST_SIZE)
+    math(EXPR MAX_INDEX ${LIST_SIZE}-1)
+
+    ## get the first output to make the output dir
+    list(GET ${OUTPUT_LIST} 0 CUR_OUTPUT)
+
+    ## create dir prefix
+    string(REGEX MATCH build/.*/ OUTPUT_PATH ${CUR_OUTPUT})
+    file(MAKE_DIRECTORY ${OUTPUT_PATH})
+
+    foreach(INDEX RANGE 0 ${MAX_INDEX}-1)
+        list(GET ${INPUT_LIST} ${INDEX} CUR_INPUT)
+        list(GET ${OUTPUT_LIST} ${INDEX} CUR_OUTPUT)
+
+
+        add_custom_command(
+            OUTPUT ${CUR_OUTPUT}
+            COMMAND protoc -I. --cpp_out=. --python_out=. build/${CUR_INPUT}
+            DEPENDS ${CUR_INPUT} ${OUTPUT_PATH} copy_proto_files_to_build
+            )
+    endforeach()
+
+endmacro(c_proto_buff)
+
+add_custom_target(copy_proto_files_to_build
+    ## This command will only works on linux, cp on macos doesn't provide --parents option
+    COMMAND cd ${SOURCE} cp --parents ${CORE_PROTO} ${PROTOCOL_PROTO} ${TREE_PROTO} ${SERVER_PROTO} ${STORAGE_PROTO} ./build
+    DEPENDS ${CORE_PROTO} ${PROTOCOL_PROTO} ${TREE_PROTO} ${SERVER_PROTO} ${STORAGE_PROTO}
+    )
+
+c_proto_buff(CORE_PROTO CORE_PROTO_SRC)
+c_proto_buff(PROTOCOL_PROTO PROTOCOL_SRC)
+c_proto_buff(TREE_PROTO TREE_PROTO_SRC)
+c_proto_buff(SERVER_PROTO SERVER_PROTO_SRC)
+c_proto_buff(STORAGE_PROTO STORAGE_PROTO_SRC)
+
+## append compiler flags (and generate a object)
+
+# logcabin srcs
+set_property(
+    SOURCE
+    Storage/Tool.cc
+    Server/Main.cc
+    ${CORE_SRC}
+    ${CLIENT_SRC}
+    ${EVENT_SRC}
+    ${RPC_SRC}
+    ${SERVER_SRC}
+    ${STORAGE_SRC}
+    ${TREE_SRC}
+    ${EXAMPLE_FILES}
+    APPEND_STRING
+    PROPERTY
+    COMPILE_FLAGS
+    " ${CMAKE_LOGCABIN_CXX_FLAGS}"
+    )
+
+# protobuf flags
+set_property(
+    SOURCE
+    ${CORE_PROTO_SRC}
+    ${PROTOCOL_SRC}
+    ${TREE_PROTO_SRC}
+    ${SERVER_PROTO_SRC}
+    ${STORAGE_PROTO_SRC}
+    APPEND_STRING
+    PROPERTY
+    COMPILE_FLAGS
+    " ${CMAKE_PROTO_CXX_FLAGS}"
+    )
+
+set_property(
+    SOURCE
+    gtest/src/gtest-all.cc
+    test/TestRunner.cc
+    ${TEST_FILES}
+    ${MOCK_FILES}
+    APPEND_STRING
+    PROPERTY
+    COMPILE_FLAGS
+    # -fno-access-control allows tests to access private members
+    " ${CMAKE_GTEST_CXX_FLAGS} ${CMAKE_LOGCABIN_CXX_FLAGS} -fno-access-control"
+    )
+
+##executable daemon
+add_executable(LogCabin
+    Server/Main.cc
+    )
+
+add_executable(ServerControl
+    Client/ServerControl.cc
+    )
+
+add_library(logcabin STATIC 
+    ${CLIENT_SRC}
+    ${TREE_SRC}
+    ${TREE_PROTO_SRC}
+    ${PROTOCOL_SRC}
+    ${RPC_SRC}
+    ${EVENT_SRC}
+    ${CORE_SRC}
+    ${CORE_PROTO_SRC}
+    )
+
+add_library(ServerSrcObject STATIC 
+    ${SERVER_SRC}
+    ${SERVER_PROTO_SRC}
+    )
+add_dependencies(ServerSrcObject ${PROTOCOL_SRC})
```

Comment:
This doesn't work with cmake 3.1x + clang++ 6.0, as StorageSrcObject will be built first. Add ${PROTOCOL_SRC} to above add_library solves the issue.